### ISSUE TITLE
Dense counterpart and pretraining data for OLMo model

### ISSUE THREAD

**Issue by aditi184** [allenai/OLMoE#32]
Is the Hugging Face checkpoint at https://huggingface.co/allenai/OLMo-1B-0724-954000steps-unsharded the dense version of the OLMoE-1B-7B-0924 model? Were these two models pretrained on the same dataset? 

In case they are not the same, is it possible to get the dense counterpart of OLMoE-1B-7B-0924 in the standard OLMo format? 

Thanks so much!

**Comment by aditi184** [comment]
And if so, could you please share the metadata.json file for OLMo-1B needed for pretraining.

**Comment by Muennighoff** [comment]
i think that is a dense version that was used for upcycling experiments; it is not used for `OLMoE-1B-7B-0924` in any way

**Comment by aditi184** [comment]
Yes, I meant whether there's a dense counterpart available. I'm running some experiments where I need to compare the MoE model (OLMoE-1B-7B-0924) with a 1B-parameter dense model that’s pretrained on the same data. Specifically, I’m referring to the comparison in Figure 4 of this paper: https://arxiv.org/pdf/2409.02060

From Figure 4 and the accompanying text, it seems that both the 1.3B dense model and the 6.9B total / 1.3B active MoE model were trained with the same configuration and data for 130B tokens. Is the dense model checkpoint from that experiment available, ideally in the standard OLMo format?

Alternatively, would the OLMo-1B checkpoint (e.g., https://huggingface.co/allenai/OLMo-1B-0724-954000steps-unsharded) have a large overlap with OLMoE’s pretraining data, making it a reasonable proxy?

I do understand the experiment in the section 4.1.1, the dense model is not trained on all of pretraining data of OLMoE.

**Comment by Muennighoff** [comment]
Yes the models from Fig 4 would be great for that, maybe @aman-17 could help release them! (they're in the bucket in the wandb linked in the caption)

Very different training data for OLMo-1B and 1B-7B so not a great proxy I think.

**Comment by aditi184** [comment]
Thanks so much @Muennighoff!!

**Comment by aditi184** [comment]
Hey @aman-17 

Could you please share these checkpoints? Thanks a lot!

**Comment by aman-17** [comment]
Hey @aditi184, it's on my TODO list, was busy w other stuff. I will probably start uploading it today and share some update by Tuesday.

**Comment by aditi184** [comment]
Thanks a lot! @aman-17

**Comment by aman-17** [comment]
> Yes, I meant whether there's a dense counterpart available. I'm running some experiments where I need to compare the MoE model (OLMoE-1B-7B-0924) with a 1B-parameter dense model that’s pretrained on the same data. Specifically, I’m referring to the comparison in Figure 4 of this paper: https://arxiv.org/pdf/2409.02060
> 
> From Figure 4 and the accompanying text, it seems that both the 1.3B dense model and the 6.9B total / 1.3B active MoE model were trained with the same configuration and data for 130B tokens. Is the dense model checkpoint from that experiment available, ideally in the standard OLMo format?
> 
> Alternatively, would the OLMo-1B checkpoint (e.g., https://huggingface.co/allenai/OLMo-1B-0724-954000steps-unsharded) have a large overlap with OLMoE’s pretraining data, making it a reasonable proxy?
> 
> I do understand the experiment in the section 4.1.1, the dense model is not trained on all of pretraining data of OLMoE.

Hey @aditi184, FYI, we have released latest OLMo2 1B model yesterday, feel free to check that out on [OLMo](https://github.com/allenai/OLMo) repo.

**Comment by aditi184** [comment]
Thanks, @aman-17 ! From what I understand, OLMo2-1B uses a different tokenizer and vocabulary compared to OLMo-1B and OLMoE-1B-7B. Specifically, OLMo2-1B is trained with tokenizers/allenai_dolma2.json (vocab size: 100,278), whereas OLMo-1B and the 1B-7B MoE model use tokenizers/allenai_gpt-neox-olmo-dolma-v1_5.json (vocab size: 50,280).

At the moment, I’m looking for a dense model (1B) that uses the same tokenizer and same or at least very similar training data as OLMoE-1B-7B, to ensure a fair comparison in my experiments.

From [here](https://huggingface.co/allenai/OLMo-1B), it looks like the OLMo's previous version (OLMo1 maybe? ) was trained using DOLMA tokens, kind of same as 1b-7b.

**Comment by aman-17** [comment]
Hey @aditi184 which checkpoints (I mean steps) do you need, I can fastforward that, and do you need unsharded format?

**Comment by aditi184** [comment]
Hey @aman-17 

I think the chechpoints I am looking for are the one around 30000th step from [this](https://wandb.ai/ai2-llm/olmoe/reports/Plot-MoE-vs-Dense--Vmlldzo4OTM0Mjkx) experiment where CE loss is minimum for both dense and MoE model. It would be great to have both sharded and unsharded versions of these checkpoints, along with their respective config.yaml and metadata.json files, so I can use them directly for continual pretraining using the OLMo repo.

For your reference, these one:

![Image](https://github.com/user-attachments/assets/d0b1f9a4-461f-4fdd-b677-2fab7707f3fa)

Could you please also share the metadata.json file for the checkpoint here: https://huggingface.co/allenai/OLMo-1B-0724-954000steps-unsharded/tree/main. It looks like this checkpoint has been trained with a large data overlap (dolma dataset) of 1b-7b. 

Thank you so much for all the help!

**Comment by aman-17** [comment]
Heyy, these look like ablation runs(not 100% sure), and we don't release them. I will cross check these internally.

**Comment by Muennighoff** [comment]
@aman-17 i think the ckpts are just lying on s3 so why can't we release them?

**Comment by aman-17** [comment]
Hey @aditi184, uploaded the checkpoints.
Links:
https://olmo-checkpoints.org/ai2-llm/olmoe-8x1b-newhp-newds-cx5-fine1/olmoe-8x1b-newhp-newds-cx5-fine1/step5000/
https://olmo-checkpoints.org/ai2-llm/olmoe-8x1b-newhp-newds-cx5-fine1/olmoe-8x1b-newhp-newds-cx5-fine1/step10000/
https://olmo-checkpoints.org/ai2-llm/olmoe-8x1b-newhp-newds-cx5-fine1/olmoe-8x1b-newhp-newds-cx5-fine1/step15000/
https://olmo-checkpoints.org/ai2-llm/olmoe-8x1b-newhp-newds-cx5-fine1/olmoe-8x1b-newhp-newds-cx5-fine1/step20000/
https://olmo-checkpoints.org/ai2-llm/olmoe-8x1b-newhp-newds-cx5-fine1/olmoe-8x1b-newhp-newds-cx5-fine1/step25000/
https://olmo-checkpoints.org/ai2-llm/olmoe-8x1b-newhp-newds-cx5-fine1/olmoe-8x1b-newhp-newds-cx5-fine1/step30000/
https://olmo-checkpoints.org/ai2-llm/olmoe-8x1b-newhp-newds-cx5-fine1/olmoe-8x1b-newhp-newds-cx5-fine1/step30993/

https://olmo-checkpoints.org/ai2-llm/olmo1b-newhp-newds-cx5/newhp-newds-cx5/step5000/
https://olmo-checkpoints.org/ai2-llm/olmo1b-newhp-newds-cx5/newhp-newds-cx5/step10000/
https://olmo-checkpoints.org/ai2-llm/olmo1b-newhp-newds-cx5/newhp-newds-cx5/step15000/
https://olmo-checkpoints.org/ai2-llm/olmo1b-newhp-newds-cx5/newhp-newds-cx5/step20000/
https://olmo-checkpoints.org/ai2-llm/olmo1b-newhp-newds-cx5/newhp-newds-cx5/step25000/
https://olmo-checkpoints.org/ai2-llm/olmo1b-newhp-newds-cx5/newhp-newds-cx5/step30000/
https://olmo-checkpoints.org/ai2-llm/olmo1b-newhp-newds-cx5/newhp-newds-cx5/step30993/


In this OLMo-1B-0724 checkpoint, you can find metadata in optim and model folders: https://olmo-checkpoints.org/ai2-llm/olmo1b-0724/step954000/

**Comment by aditi184** [comment]
Thank you so much for this!!
This is really helpful!

**Comment by aditi184** [comment]
Hi @aman-17 

When I click on these links I get the error: Error 404
Object not found
This object does not exist or is not publicly accessible at this URL. Check the URL of the object that you're looking for or contact the owner to enable Public access.

and when I try downloading via wget, I get similar error. 

Is there another way to download these or can you enable the public access on this bucket?

**Comment by aman-17** [comment]
Hey @aditi184, they are public. Use boto3 to download these :)

**Comment by aditi184** [comment]
I think I might need to more details to download using boto3, like the bucket_name and key and if I can download as an UNSIGNED USER. 

```
import boto3
from botocore import UNSIGNED
from botocore.config import Config
import os

s3 = boto3.client('s3', config=Config(signature_version=UNSIGNED)) #NOT SURE

bucket_name = 'ai2-llm' # NOT SURE

object_keys = [
    'olmoe-8x1b-newhp-newds-cx5-fine1/olmoe-8x1b-newhp-newds-cx5-fine1/step5000', #NOT SURE
]

download_dir = 'checkpoints'

for key in object_keys:
    local_path = os.path.join(download_dir, os.path.basename(key))
    os.makedirs(os.path.dirname(local_path), exist_ok=True)

    s3.download_file(bucket_name, key, local_path)
    print(f'Downloaded {key} to {local_path}')
```

**Comment by aman-17** [comment]
No, you don't need those. Instead, you can try this approach, easier to debug: https://github.com/allenai/OLMo/issues/773#issuecomment-2569624139

**Comment by dirkgr** [comment]
The problem is that the HTTP links don't allow listing directories. HTTP works when you know the exact filename. @aman-17, can you give @aditi184 an example of which file names are under each of those checkpoint paths?

**Comment by aditi184** [comment]
Hi @dirkgr, thanks for pointing that out.

@aman-17 
I had initially guessed the folder and file names by trying different directory and file names for unsharded/sharded checkpoints, as well as the model rank, and it turns out these are sharded checkpoints with rank 8. I was able to successfully download and load the MoE model.

However, I'm facing issues when loading the dense version. There seems to be a config mismatch as I'm getting very high perplexity scores (30k, 50k), which suggests the model might be getting randomly initialized.

Also, I'm a bit confused about the config uploaded for the dense model. It includes the following MoE-related fields:
 ```
  moe_num_experts: 8
  moe_top_k: 2
  moe_mlp_impl: sparse
  moe_log_expert_assignment: true
  moe_shared_expert: false
  moe_lbl_in_fp32: false
  moe_loss_weight: 0.1
  moe_zloss_weight: null
  moe_dropless: true
  moe_capacity_factor: 1.25
```

It would be helpful if you could clarify the purpose of these settings in the dense config.

I think I might have made a mistake in downloading the dense model files. Could you please share the exact names of the files and folders under the checkpoint paths?

**Comment by aman-17** [comment]
For `olmo1b-newhp-newds-cx5` checkpoints, there will be one config.yaml file and three folders: optim, model, and train. Optim and model contain metadata.json .safetensors files and train has .pt files. These .safetensors and .pt are named as rank_1, rank_2...

Can I ask the reason for comparing these two models? The data is so different that I’m not sure what conclusions you’ll be able to draw from it.

**Comment by aditi184** [comment]
Thanks for the filenames, @aman-17.

These checkpoints are from the ablation study discussed in [Section 4.1.1 of the paper](https://arxiv.org/pdf/2409.02060#page=7.81) on Mixture-of-Experts vs. Dense. In the caption of Figure 4, it mentions: "Apart from MoE-related changes, we train both with the same configuration for 130B tokens." The text in that section also says "we compare MoEs and dense models in a controlled setup", which led me to believe that aside from the architecture, all other settings, including data, were the same, and hence the models could be meaningfully compared.

I'm training MoEs on multilingual data to study some properties and was hoping to begin with checkpoints where MoE and dense models were trained under the same setup, including the data. Am I misunderstanding something here?

**Comment by aditi184** [comment]
Also, just checked in the configs of both of them, the data looks the same.
